src/core/Akka/Actor/FSM.cs

```diff
@@ -1137,43 +1235,38 @@ private string GetSourceString(object source)
             ApplyState(upcomingState);
         }
 
-        private void ApplyState(State<TState, TData> upcomingState)
+        private void ApplyState(State<TState, TData> nextState)
         {
-            if (upcomingState.StopReason == null){ MakeTransition(upcomingState);
-                return;
+            if (nextState.StopReason == null)
+            {
+                MakeTransition(nextState);
             }
-            var replies = upcomingState.Replies;
-            replies.Reverse();
-            foreach (var reply in replies)
+            else
             {
-                Sender.Tell(reply);
+                nextState.Replies.Reverse().ForEach(r => Sender.Tell(r));
+                Terminate(nextState);
+                Context.Stop(Self);
             }
-            Terminate(upcomingState);
-            Context.Stop(Self);
         }
 
-        private void MakeTransition(State<TState, TData> upcomingState)
+        private void MakeTransition(State<TState, TData> nextState)
         {
-            if (!_stateFunctions.ContainsKey(upcomingState.StateName))
+            if (!_stateFunctions.ContainsKey(nextState.StateName))
             {
-                Terminate(
-                    Stay()
-                        .WithStopReason(
-                            new Failure(String.Format("Next state {0} does not exist", upcomingState.StateName))));
+                Terminate(Stay().WithStopReason(new Failure($"Next state {nextState.StateName} does not exist")));
             }
             else
             {
-                var replies = upcomingState.Replies;
-                replies.Reverse();
-                foreach (var r in replies) { Sender.Tell(r); }
-                if (!_currentState.StateName.Equals(upcomingState.StateName))
+                nextState.Replies.Reverse().ForEach(r => Sender.Tell(r));
```

Comment:
instead of reversing the list and then iterating, why not do both like

```csharp
for (int i = nextState.Replies.Count - 1; i >= 0; i--)
{
    Sender.Tell(nextState.Replies[i]);
}
```